Allow (temporarily?) non-fake input during ONNX export with fake mode

thiagocrepaldi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #105477
* #105247
* __->__ #105246

Although input and model are expected to be fake during ONNX export with fake mode enabled, apparently some models can create new parameters during tracing. That makes internal checks on dynamo side to fail when we dont set `allow_non_fake_input=True` for `torch._dynamo.export`.

https://github.com/pytorch/pytorch/issues/105077 tracks this issue and if a proper fix is done, we will set `allow_non_fake_input=False` again

Additionally to that, a possible bug was found at torch.nn.Module.state_dict() in which some registered buffers are not listed.

This is being tracked by https://github.com/pytorch/pytorch/issues/105233 but in the mean time, we are merging `state_dict()` and `named_buffers()` results to create a full `state_dict` for the model

Two more complex/larger tests are added to the ONNX export which are the same for the experimental symbolic tracing: tiny gpt2 and toy mlp (https://github.com/pytorch/pytorch/blob/main/test/onnx/test_fx_to_onnx_with_onnxruntime.py#L766-L825)

ps: https://github.com/pytorch/pytorch/issues/105464 tracks pending tasks/limitations from this PR

Commit:
Allow (temporarily?) non-fake input during ONNX export with fake mode

Although input and model are expected to be fake during ONNX export with
fake mode enabled, apparently some models can create new parameters
during tracing. That makes internal checks on dynamo side to fail when
we dont set `allow_non_fake_input=True` for `torch._dynamo.export`.

https://github.com/pytorch/pytorch/issues/105077 tracks this issue and
if a proper fix is done, we will set `allow_non_fake_input=False` again

A possible bug was found at torch.nn.Module.state_dict() in which some
registered buffers are not listed.

This is being tracked by https://github.com/pytorch/pytorch/issues/105233
but in the mean time, we are merging state_dict() and named_buffers()
results to create a full state_dict for the model

Two more complex/larger tests are added to the ONNX export with symbolic
tracing: tiny gpt2 and toy mlp

[ghstack-poisoned]

Commit:
Update on "Allow (temporarily?) non-fake input during ONNX export with fake mode"


Although input and model are expected to be fake during ONNX export with fake mode enabled, apparently some models can create new parameters during tracing. That makes internal checks on dynamo side to fail when we dont set `allow_non_fake_input=True` for `torch._dynamo.export`.

https://github.com/pytorch/pytorch/issues/105077 tracks this issue and if a proper fix is done, we will set `allow_non_fake_input=False` again

Additionally to that, a possible bug was found at torch.nn.Module.state_dict() in which some registered buffers are not listed.

This is being tracked by https://github.com/pytorch/pytorch/issues/105233 but in the mean time, we are merging `state_dict()` and `named_buffers()` results to create a full `state_dict` for the model

Two more complex/larger tests are added to the ONNX export with symbolic tracing: tiny gpt2 and toy mlp

[ghstack-poisoned]

Commit:
Update on "Allow (temporarily?) non-fake input during ONNX export with fake mode"


Although input and model are expected to be fake during ONNX export with fake mode enabled, apparently some models can create new parameters during tracing. That makes internal checks on dynamo side to fail when we dont set `allow_non_fake_input=True` for `torch._dynamo.export`.

https://github.com/pytorch/pytorch/issues/105077 tracks this issue and if a proper fix is done, we will set `allow_non_fake_input=False` again

Additionally to that, a possible bug was found at torch.nn.Module.state_dict() in which some registered buffers are not listed.

This is being tracked by https://github.com/pytorch/pytorch/issues/105233 but in the mean time, we are merging `state_dict()` and `named_buffers()` results to create a full `state_dict` for the model

Two more complex/larger tests are added to the ONNX export with symbolic tracing: tiny gpt2 and toy mlp

[ghstack-poisoned]

Commit:
Update on "Allow (temporarily?) non-fake input during ONNX export with fake mode"


Although input and model are expected to be fake during ONNX export with fake mode enabled, apparently some models can create new parameters during tracing. That makes internal checks on dynamo side to fail when we dont set `allow_non_fake_input=True` for `torch._dynamo.export`.

https://github.com/pytorch/pytorch/issues/105077 tracks this issue and if a proper fix is done, we will set `allow_non_fake_input=False` again

Additionally to that, a possible bug was found at torch.nn.Module.state_dict() in which some registered buffers are not listed.

This is being tracked by https://github.com/pytorch/pytorch/issues/105233 but in the mean time, we are merging `state_dict()` and `named_buffers()` results to create a full `state_dict` for the model

Two more complex/larger tests are added to the ONNX export with symbolic tracing: tiny gpt2 and toy mlp

[ghstack-poisoned]

Commit:
Update on "Allow (temporarily?) non-fake input during ONNX export with fake mode"


Although input and model are expected to be fake during ONNX export with fake mode enabled, apparently some models can create new parameters during tracing. That makes internal checks on dynamo side to fail when we dont set `allow_non_fake_input=True` for `torch._dynamo.export`.

https://github.com/pytorch/pytorch/issues/105077 tracks this issue and if a proper fix is done, we will set `allow_non_fake_input=False` again

Additionally to that, a possible bug was found at torch.nn.Module.state_dict() in which some registered buffers are not listed.

This is being tracked by https://github.com/pytorch/pytorch/issues/105233 but in the mean time, we are merging `state_dict()` and `named_buffers()` results to create a full `state_dict` for the model

Two more complex/larger tests are added to the ONNX export with symbolic tracing: tiny gpt2 and toy mlp

[ghstack-poisoned]

thiagocrepaldi commented on test/onnx/test_fx_to_onnx.py:
```diff
@@ -361,11 +361,13 @@ def forward(self, x):
-            _ = torch.onnx.dynamo_export(
-                real_model, fake_x, export_options=export_options
-            )
+        # TODO: Delete the test below if https://github.com/pytorch/pytorch/pull/105246
```
Tracked by https://github.com/pytorch/pytorch/issues/105464 subtask

Commit:
Update on "Allow (temporarily?) non-fake input during ONNX export with fake mode"


Although input and model are expected to be fake during ONNX export with fake mode enabled, apparently some models can create new parameters during tracing. That makes internal checks on dynamo side to fail when we dont set `allow_non_fake_input=True` for `torch._dynamo.export`.

https://github.com/pytorch/pytorch/issues/105077 tracks this issue and if a proper fix is done, we will set `allow_non_fake_input=False` again

Additionally to that, a possible bug was found at torch.nn.Module.state_dict() in which some registered buffers are not listed.

This is being tracked by https://github.com/pytorch/pytorch/issues/105233 but in the mean time, we are merging `state_dict()` and `named_buffers()` results to create a full `state_dict` for the model

Two more complex/larger tests are added to the ONNX export with symbolic tracing: tiny gpt2 and toy mlp

[ghstack-poisoned]

titaiwangms commented on test/onnx/test_fx_to_onnx_with_onnxruntime.py:
```diff
@@ -932,6 +941,70 @@ def create_pytorch_only_extra_kwargs():
         )
 
+    @pytorch_test_common.skip_op_level_debug_test(
+        "op_level_debug_test does not support FakeTensor yet."
```
What is the error we have here? op_level_debug is randomly generating Tensors based on FakeTensor. This kind of doesn't make sense to me.

thiagocrepaldi replied:
The current code only generates *inputs* based on `FakeTensor`, but now the model parameters and buffers are also fake and probably not being replaced by actual tensors
```bash
========================================================================== test session starts ==========================================================================
platform linux -- Python 3.8.16, pytest-7.4.0, pluggy-1.0.0 -- /opt/conda/envs/ptca/bin/python
cachedir: .pytest_cache
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/opt/pytorch/.hypothesis/examples')
Using --randomly-seed=3200217290
benchmark: 4.0.0 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
rootdir: /opt/pytorch
configfile: pytest.ini
plugins: mpi-0.6, xdoctest-1.1.1, hypothesis-6.79.2, randomly-3.12.0, hydra-core-1.3.2, anyio-3.7.0, xdist-3.3.1, nbval-0.10.0, benchmark-4.0.0, subtests-0.11.0, cov-4.1.0
collecting ... No CUDA runtime is found, using CUDA_HOME='/usr/local/cuda'
/opt/pytorch/torch/__init__.py:615: UserWarning: torch.set_default_tensor_type() is deprecated as of PyTorch 2.1, please use torch.set_default_dtype() and torch.set_default_device() as alternatives. (Triggered internally at /opt/pytorch/torch/csrc/tensor/python_tensor.cpp:451.)
  _C._set_default_tensor_type(t)
/opt/conda/envs/ptca/lib/python3.8/site-packages/torchvision/io/image.py:13: UserWarning: Failed to load image Python extension: 'libc10_cuda.so: cannot open shared object file: No such file or directory'If you don't plan on using image functionality from `torchvision.io`, you can ignore this warning. Otherwise, there might be something wrong with your environment. Did you have `libjpeg` or `libpng` installed before building `torchvision` from source?
  warn(
collected 1 item                                                                                                                                                        

test/onnx/test_fx_to_onnx_with_onnxruntime.py::TestFxToOnnxFakeTensorWithOnnxRuntime_op_level_debug_True_dynamic_shapes_False_load_checkpoint_during_init_False_export_within_fake_mode_False::test_fake_tensor_mode_simple /opt/pytorch/torch/overrides.py:111: UserWarning: 'has_cuda' is deprecated, please use 'torch.backends.cuda.is_built()'
  torch.has_cuda,
/opt/pytorch/torch/overrides.py:112: UserWarning: 'has_cudnn' is deprecated, please use 'torch.backends.cudnn.is_available()'
  torch.has_cudnn,
/opt/pytorch/torch/overrides.py:118: UserWarning: 'has_mps' is deprecated, please use 'torch.backends.mps.is_built()'
  torch.has_mps,
/opt/pytorch/torch/overrides.py:119: UserWarning: 'has_mkldnn' is deprecated, please use 'torch.backends.mkldnn.is_available()'
  torch.has_mkldnn,
[2023-07-18 19:25:52,968] torch.fx.experimental.symbolic_shapes: [WARNING] Ignored guard Eq(s1, 2) == True, this could result in accuracy problems
[2023-07-18 19:25:53,275] torch.onnx: [ERROR] FX Node: call_function:aten.view.default[name=view]. Raised from:
[2023-07-18 19:25:53,275] torch.onnx: [ERROR]     RuntimeError: /opt/pytorch/build/aten/src/ATen/RegisterCompositeExplicitAutograd.cpp:3357: SymIntArrayRef expected to contain only concrete integers
[2023-07-18 19:25:53,275] torch.onnx: [ERROR] ### PyTorch source information
[2023-07-18 19:25:53,275] torch.onnx: [ERROR] ```
[2023-07-18 19:25:53,275] torch.onnx: [ERROR]   File "/opt/pytorch/test/onnx/test_fx_to_onnx_with_onnxruntime.py", line 973, in forward
[2023-07-18 19:25:53,275] torch.onnx: [ERROR]     out = self.linear(x)
[2023-07-18 19:25:53,275] torch.onnx: [ERROR] 
[2023-07-18 19:25:53,275] torch.onnx: [ERROR] ```
[2023-07-18 19:25:53,275] torch.onnx: [ERROR] ### Function Signature FxOnnxInterpreter.run_node
[2023-07-18 19:25:53,275] torch.onnx: [ERROR] - self: <class 'torch.onnx._internal.fx.fx_onnx_interpreter.FxOnnxInterpreter'>
[2023-07-18 19:25:53,275] torch.onnx: [ERROR] - node: fx.Node(view[call_function]'aten.view.default')
[2023-07-18 19:25:53,275] torch.onnx: [ERROR] - fx_graph_module: torch.fx.GraphModule(<lambda>)
[2023-07-18 19:25:53,275] torch.onnx: [ERROR] - onnxfunction_dispatcher: <class 'torch.onnx._internal.fx.onnxfunction_dispatcher.OnnxFunctionDispatcher'>
[2023-07-18 19:25:53,275] torch.onnx: [ERROR] - op_level_debug: <class 'bool'>
[2023-07-18 19:25:53,275] torch.onnx: [ERROR] - onnxscript_graph: <class 'onnxscript.function_libs.torch_lib.graph_building.TorchScriptGraph'>
[2023-07-18 19:25:53,275] torch.onnx: [ERROR] - onnxscript_tracer: <class 'onnxscript.function_libs.torch_lib.graph_building.TorchScriptTracingEvaluator'>
[2023-07-18 19:25:53,275] torch.onnx: [ERROR] - fx_name_to_onnxscript_value: <class 'dict'>
[2023-07-18 19:25:53,275] torch.onnx: [ERROR] ### Exception log
[2023-07-18 19:25:53,275] torch.onnx: [ERROR] ```
[2023-07-18 19:25:53,275] torch.onnx: [ERROR] Traceback (most recent call last):
[2023-07-18 19:25:53,275] torch.onnx: [ERROR] 
[2023-07-18 19:25:53,275] torch.onnx: [ERROR]   File "/opt/pytorch/torch/onnx/_internal/diagnostics/infra/decorator.py", line 127, in wrapper
[2023-07-18 19:25:53,275] torch.onnx: [ERROR]     return_values = fn(*args, **kwargs)
[2023-07-18 19:25:53,275] torch.onnx: [ERROR] 
[2023-07-18 19:25:53,275] torch.onnx: [ERROR]   File "/opt/pytorch/torch/onnx/_internal/fx/fx_onnx_interpreter.py", line 395, in run_node
[2023-07-18 19:25:53,275] torch.onnx: [ERROR]     self.call_function(
[2023-07-18 19:25:53,275] torch.onnx: [ERROR] 
[2023-07-18 19:25:53,275] torch.onnx: [ERROR]   File "<@beartype(torch.onnx._internal.fx.fx_onnx_interpreter.FxOnnxInterpreter.call_function) at 0x7f1ac157c670>", line 113, in call_function
[2023-07-18 19:25:53,275] torch.onnx: [ERROR] 
[2023-07-18 19:25:53,275] torch.onnx: [ERROR]   File "/opt/pytorch/torch/onnx/_internal/fx/fx_onnx_interpreter.py", line 583, in call_function
[2023-07-18 19:25:53,275] torch.onnx: [ERROR]     torch_args, torch_kwargs = op_validation.wrap_fx_args_as_torch_args(
[2023-07-18 19:25:53,275] torch.onnx: [ERROR] 
[2023-07-18 19:25:53,275] torch.onnx: [ERROR]   File "<@beartype(torch.onnx._internal.fx.op_validation.wrap_fx_args_as_torch_args) at 0x7f1ac155e550>", line 70, in wrap_fx_args_as_torch_args
[2023-07-18 19:25:53,275] torch.onnx: [ERROR] 
[2023-07-18 19:25:53,275] torch.onnx: [ERROR]   File "/opt/pytorch/torch/onnx/_internal/fx/op_validation.py", line 250, in wrap_fx_args_as_torch_args
[2023-07-18 19:25:53,275] torch.onnx: [ERROR]     torch_args: List[fx_type_utils.Argument] = _fx_args_to_torch_args(complete_args)
[2023-07-18 19:25:53,275] torch.onnx: [ERROR] 
[2023-07-18 19:25:53,275] torch.onnx: [ERROR]   File "<@beartype(torch.onnx._internal.fx.op_validation._fx_args_to_torch_args) at 0x7f1ac155e3a0>", line 51, in _fx_args_to_torch_args
[2023-07-18 19:25:53,275] torch.onnx: [ERROR] 
[2023-07-18 19:25:53,275] torch.onnx: [ERROR]   File "/opt/pytorch/torch/onnx/_internal/fx/op_validation.py", line 212, in _fx_args_to_torch_args
[2023-07-18 19:25:53,275] torch.onnx: [ERROR]     real_tensor = generate_random_tensors(
[2023-07-18 19:25:53,275] torch.onnx: [ERROR] 
[2023-07-18 19:25:53,275] torch.onnx: [ERROR]   File "<@beartype(torch.onnx._internal.fx.op_validation.generate_random_tensors) at 0x7f1ac155e1f0>", line 52, in generate_random_tensors
[2023-07-18 19:25:53,275] torch.onnx: [ERROR] 
[2023-07-18 19:25:53,275] torch.onnx: [ERROR]   File "/opt/pytorch/torch/onnx/_internal/fx/op_validation.py", line 195, in generate_random_tensors
[2023-07-18 19:25:53,275] torch.onnx: [ERROR]     return torch.randn(shape, dtype=dtype)
[2023-07-18 19:25:53,275] torch.onnx: [ERROR] 
[2023-07-18 19:25:53,275] torch.onnx: [ERROR] RuntimeError: /opt/pytorch/build/aten/src/ATen/RegisterCompositeExplicitAutograd.cpp:3357: SymIntArrayRef expected to contain only concrete integers
[2023-07-18 19:25:53,275] torch.onnx: [ERROR] 
[2023-07-18 19:25:53,275] torch.onnx: [ERROR] ```
[2023-07-18 19:25:53,276] torch.onnx: [ERROR] FxOnnxInterpreter.run. Raised from:
[2023-07-18 19:25:53,276] torch.onnx: [ERROR]     RuntimeErrorWithDiagnostic: FX Node: call_function:aten.view.default[name=view]. Raised from:
[2023-07-18 19:25:53,276] torch.onnx: [ERROR]     RuntimeError: /opt/pytorch/build/aten/src/ATen/RegisterCompositeExplicitAutograd.cpp:3357: SymIntArrayRef expected to contain only concrete integers
[2023-07-18 19:25:53,276] torch.onnx: [ERROR] ### Function Signature FxOnnxInterpreter.run
[2023-07-18 19:25:53,276] torch.onnx: [ERROR] - self: <class 'torch.onnx._internal.fx.fx_onnx_interpreter.FxOnnxInterpreter'>
[2023-07-18 19:25:53,276] torch.onnx: [ERROR] - fx_graph_module: torch.fx.GraphModule(<lambda>)
[2023-07-18 19:25:53,276] torch.onnx: [ERROR] - onnxfunction_dispatcher: <class 'torch.onnx._internal.fx.onnxfunction_dispatcher.OnnxFunctionDispatcher'>
[2023-07-18 19:25:53,276] torch.onnx: [ERROR] - op_level_debug: <class 'bool'>
[2023-07-18 19:25:53,276] torch.onnx: [ERROR] ### Exception log
[2023-07-18 19:25:53,276] torch.onnx: [ERROR] ```
[2023-07-18 19:25:53,276] torch.onnx: [ERROR] Traceback (most recent call last):
[2023-07-18 19:25:53,276] torch.onnx: [ERROR] 
[2023-07-18 19:25:53,276] torch.onnx: [ERROR]   File "/opt/pytorch/torch/onnx/_internal/diagnostics/infra/decorator.py", line 127, in wrapper
[2023-07-18 19:25:53,276] torch.onnx: [ERROR]     return_values = fn(*args, **kwargs)
[2023-07-18 19:25:53,276] torch.onnx: [ERROR] 
[2023-07-18 19:25:53,276] torch.onnx: [ERROR]   File "/opt/pytorch/torch/onnx/_internal/fx/fx_onnx_interpreter.py", line 395, in run_node
[2023-07-18 19:25:53,276] torch.onnx: [ERROR]     self.call_function(
[2023-07-18 19:25:53,276] torch.onnx: [ERROR] 
[2023-07-18 19:25:53,276] torch.onnx: [ERROR]   File "<@beartype(torch.onnx._internal.fx.fx_onnx_interpreter.FxOnnxInterpreter.call_function) at 0x7f1ac157c670>", line 113, in call_function
[2023-07-18 19:25:53,276] torch.onnx: [ERROR] 
[2023-07-18 19:25:53,276] torch.onnx: [ERROR]   File "/opt/pytorch/torch/onnx/_internal/fx/fx_onnx_interpreter.py", line 583, in call_function
[2023-07-18 19:25:53,276] torch.onnx: [ERROR]     torch_args, torch_kwargs = op_validation.wrap_fx_args_as_torch_args(
[2023-07-18 19:25:53,276] torch.onnx: [ERROR] 
[2023-07-18 19:25:53,276] torch.onnx: [ERROR]   File "<@beartype(torch.onnx._internal.fx.op_validation.wrap_fx_args_as_torch_args) at 0x7f1ac155e550>", line 70, in wrap_fx_args_as_torch_args
[2023-07-18 19:25:53,276] torch.onnx: [ERROR] 
[2023-07-18 19:25:53,276] torch.onnx: [ERROR]   File "/opt/pytorch/torch/onnx/_internal/fx/op_validation.py", line 250, in wrap_fx_args_as_torch_args
[2023-07-18 19:25:53,276] torch.onnx: [ERROR]     torch_args: List[fx_type_utils.Argument] = _fx_args_to_torch_args(complete_args)
[2023-07-18 19:25:53,276] torch.onnx: [ERROR] 
[2023-07-18 19:25:53,276] torch.onnx: [ERROR]   File "<@beartype(torch.onnx._internal.fx.op_validation._fx_args_to_torch_args) at 0x7f1ac155e3a0>", line 51, in _fx_args_to_torch_args
[2023-07-18 19:25:53,276] torch.onnx: [ERROR] 
[2023-07-18 19:25:53,276] torch.onnx: [ERROR]   File "/opt/pytorch/torch/onnx/_internal/fx/op_validation.py", line 212, in _fx_args_to_torch_args
[2023-07-18 19:25:53,276] torch.onnx: [ERROR]     real_tensor = generate_random_tensors(
[2023-07-18 19:25:53,276] torch.onnx: [ERROR] 
[2023-07-18 19:25:53,276] torch.onnx: [ERROR]   File "<@beartype(torch.onnx._internal.fx.op_validation.generate_random_tensors) at 0x7f1ac155e1f0>", line 52, in generate_random_tensors
[2023-07-18 19:25:53,276] torch.onnx: [ERROR] 
[2023-07-18 19:25:53,276] torch.onnx: [ERROR]   File "/opt/pytorch/torch/onnx/_internal/fx/op_validation.py", line 195, in generate_random_tensors
[2023-07-18 19:25:53,276] torch.onnx: [ERROR]     return torch.randn(shape, dtype=dtype)
[2023-07-18 19:25:53,276] torch.onnx: [ERROR] 
[2023-07-18 19:25:53,276] torch.onnx: [ERROR] RuntimeError: /opt/pytorch/build/aten/src/ATen/RegisterCompositeExplicitAutograd.cpp:3357: SymIntArrayRef expected to contain only concrete integers
[2023-07-18 19:25:53,276] torch.onnx: [ERROR] 
[2023-07-18 19:25:53,276] torch.onnx: [ERROR] 
[2023-07-18 19:25:53,276] torch.onnx: [ERROR] The above exception was the direct cause of the following exception:
[2023-07-18 19:25:53,276] torch.onnx: [ERROR] 
[2023-07-18 19:25:53,276] torch.onnx: [ERROR] 
[2023-07-18 19:25:53,276] torch.onnx: [ERROR] Traceback (most recent call last):
[2023-07-18 19:25:53,276] torch.onnx: [ERROR] 
[2023-07-18 19:25:53,276] torch.onnx: [ERROR]   File "/opt/pytorch/torch/onnx/_internal/diagnostics/infra/decorator.py", line 127, in wrapper
[2023-07-18 19:25:53,276] torch.onnx: [ERROR]     return_values = fn(*args, **kwargs)
[2023-07-18 19:25:53,276] torch.onnx: [ERROR] 
[2023-07-18 19:25:53,276] torch.onnx: [ERROR]   File "/opt/pytorch/torch/onnx/_internal/fx/fx_onnx_interpreter.py", line 454, in run
[2023-07-18 19:25:53,276] torch.onnx: [ERROR]     self.run_node(
[2023-07-18 19:25:53,276] torch.onnx: [ERROR] 
[2023-07-18 19:25:53,276] torch.onnx: [ERROR]   File "<@beartype(torch.onnx._internal.fx.fx_onnx_interpreter.FxOnnxInterpreter.run_node) at 0x7f1ac157c310>", line 133, in run_node
[2023-07-18 19:25:53,276] torch.onnx: [ERROR] 
[2023-07-18 19:25:53,276] torch.onnx: [ERROR]   File "/opt/pytorch/torch/onnx/_internal/diagnostics/infra/decorator.py", line 142, in wrapper
[2023-07-18 19:25:53,276] torch.onnx: [ERROR]     ctx.log_and_raise_if_error(diag)
[2023-07-18 19:25:53,276] torch.onnx: [ERROR] 
[2023-07-18 19:25:53,276] torch.onnx: [ERROR]   File "/opt/pytorch/torch/onnx/_internal/diagnostics/infra/context.py", line 265, in log_and_raise_if_error
[2023-07-18 19:25:53,276] torch.onnx: [ERROR]     raise RuntimeErrorWithDiagnostic(
[2023-07-18 19:25:53,276] torch.onnx: [ERROR] 
[2023-07-18 19:25:53,276] torch.onnx: [ERROR] torch.onnx._internal.diagnostics.infra.context.RuntimeErrorWithDiagnostic: FX Node: call_function:aten.view.default[name=view]. Raised from:
[2023-07-18 19:25:53,276] torch.onnx: [ERROR]     RuntimeError: /opt/pytorch/build/aten/src/ATen/RegisterCompositeExplicitAutograd.cpp:3357: SymIntArrayRef expected to contain only concrete integers
[2023-07-18 19:25:53,276] torch.onnx: [ERROR] 
[2023-07-18 19:25:53,276] torch.onnx: [ERROR] ```
FAILED [2.1101s]

=============================================================================== FAILURES ================================================================================
_ TestFxToOnnxFakeTensorWithOnnxRuntime_op_level_debug_True_dynamic_shapes_False_load_checkpoint_during_init_False_export_within_fake_mode_False.test_fake_tensor_mode_simple _
Traceback (most recent call last):
  File "/opt/pytorch/torch/onnx/_internal/diagnostics/infra/decorator.py", line 127, in wrapper
    return_values = fn(*args, **kwargs)
  File "/opt/pytorch/torch/onnx/_internal/fx/fx_onnx_interpreter.py", line 395, in run_node
    self.call_function(
  File "<@beartype(torch.onnx._internal.fx.fx_onnx_interpreter.FxOnnxInterpreter.call_function) at 0x7f1ac157c670>", line 113, in call_function
  File "/opt/pytorch/torch/onnx/_internal/fx/fx_onnx_interpreter.py", line 583, in call_function
    torch_args, torch_kwargs = op_validation.wrap_fx_args_as_torch_args(
  File "<@beartype(torch.onnx._internal.fx.op_validation.wrap_fx_args_as_torch_args) at 0x7f1ac155e550>", line 70, in wrap_fx_args_as_torch_args
  File "/opt/pytorch/torch/onnx/_internal/fx/op_validation.py", line 250, in wrap_fx_args_as_torch_args
    torch_args: List[fx_type_utils.Argument] = _fx_args_to_torch_args(complete_args)
  File "<@beartype(torch.onnx._internal.fx.op_validation._fx_args_to_torch_args) at 0x7f1ac155e3a0>", line 51, in _fx_args_to_torch_args
  File "/opt/pytorch/torch/onnx/_internal/fx/op_validation.py", line 212, in _fx_args_to_torch_args
    real_tensor = generate_random_tensors(
  File "<@beartype(torch.onnx._internal.fx.op_validation.generate_random_tensors) at 0x7f1ac155e1f0>", line 52, in generate_random_tensors
  File "/opt/pytorch/torch/onnx/_internal/fx/op_validation.py", line 195, in generate_random_tensors
    return torch.randn(shape, dtype=dtype)
RuntimeError: /opt/pytorch/build/aten/src/ATen/RegisterCompositeExplicitAutograd.cpp:3357: SymIntArrayRef expected to contain only concrete integers

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/opt/pytorch/torch/onnx/_internal/diagnostics/infra/decorator.py", line 127, in wrapper
    return_values = fn(*args, **kwargs)
  File "/opt/pytorch/torch/onnx/_internal/fx/fx_onnx_interpreter.py", line 454, in run
    self.run_node(
  File "<@beartype(torch.onnx._internal.fx.fx_onnx_interpreter.FxOnnxInterpreter.run_node) at 0x7f1ac157c310>", line 133, in run_node
  File "/opt/pytorch/torch/onnx/_internal/diagnostics/infra/decorator.py", line 142, in wrapper
    ctx.log_and_raise_if_error(diag)
  File "/opt/pytorch/torch/onnx/_internal/diagnostics/infra/context.py", line 265, in log_and_raise_if_error
    raise RuntimeErrorWithDiagnostic(
torch.onnx._internal.diagnostics.infra.context.RuntimeErrorWithDiagnostic: FX Node: call_function:aten.view.default[name=view]. Raised from:
    RuntimeError: /opt/pytorch/build/aten/src/ATen/RegisterCompositeExplicitAutograd.cpp:3357: SymIntArrayRef expected to contain only concrete integers

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/opt/pytorch/torch/onnx/_internal/exporter.py", line 832, in dynamo_export
    return Exporter(
  File "/opt/pytorch/torch/onnx/_internal/exporter.py", line 661, in export
    onnxscript_graph = fx_interpreter.run(
  File "<@beartype(torch.onnx._internal.fx.fx_onnx_interpreter.FxOnnxInterpreter.run) at 0x7f1ac157c4c0>", line 72, in run
  File "/opt/pytorch/torch/onnx/_internal/diagnostics/infra/decorator.py", line 142, in wrapper
    ctx.log_and_raise_if_error(diag)
  File "/opt/pytorch/torch/onnx/_internal/diagnostics/infra/context.py", line 265, in log_and_raise_if_error
    raise RuntimeErrorWithDiagnostic(
torch.onnx._internal.diagnostics.infra.context.RuntimeErrorWithDiagnostic: FxOnnxInterpreter.run. Raised from:
    RuntimeErrorWithDiagnostic: FX Node: call_function:aten.view.default[name=view]. Raised from:
    RuntimeError: /opt/pytorch/build/aten/src/ATen/RegisterCompositeExplicitAutograd.cpp:3357: SymIntArrayRef expected to contain only concrete integers

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/opt/conda/envs/ptca/lib/python3.8/unittest/case.py", line 60, in testPartExecutor
    yield
  File "/opt/conda/envs/ptca/lib/python3.8/unittest/case.py", line 676, in run
    self._callTestMethod(testMethod)
  File "/opt/conda/envs/ptca/lib/python3.8/unittest/case.py", line 633, in _callTestMethod
    method()
  File "/opt/pytorch/torch/testing/_internal/common_utils.py", line 2354, in wrapper
    method(*args, **kwargs)
  File "/opt/pytorch/test/onnx/test_fx_to_onnx_with_onnxruntime.py", line 984, in test_fake_tensor_mode_simple
    self._test_fake_tensor_mode_exporter(
  File "<@beartype(test_fx_to_onnx_with_onnxruntime.TestFxToOnnxFakeTensorWithOnnxRuntime._test_fake_tensor_mode_exporter) at 0x7f1ac4c201f0>", line 127, in _test_fake_tensor_mode_exporter
  File "/opt/pytorch/test/onnx/test_fx_to_onnx_with_onnxruntime.py", line 934, in _test_fake_tensor_mode_exporter
    export_output = torch.onnx.dynamo_export(
  File "<@beartype(torch.onnx._internal.exporter.dynamo_export) at 0x7f1b630de940>", line 53, in dynamo_export
  File "/opt/pytorch/torch/onnx/_internal/exporter.py", line 847, in dynamo_export
    raise OnnxExporterError(
torch.onnx.OnnxExporterError: Failed to export the model to ONNX. Generating SARIF report at report_dynamo_export.sarif. Please report a bug on PyTorch Github: https://github.com/pytorch/pytorch/issues

To execute this test, run the following from the base repo dir:
     python test/onnx/test_fx_to_onnx_with_onnxruntime.py -k test_fake_tensor_mode_simple

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0
======================================================================== short test summary info ========================================================================
FAILED [2.1101s] test/onnx/test_fx_to_onnx_with_onnxruntime.py::TestFxToOnnxFakeTensorWithOnnxRuntime_op_level_debug_True_dynamic_shapes_False_load_checkpoint_during_init_False_export_within_fake_mode_False::test_fake_tensor_mode_simple - torch.onnx.OnnxExporterError: Failed to export the model to ONNX. Generating SARIF report at report_dynamo_export.sarif. Please report a bug on PyTorch Github: http...
=========================================================================== 1 failed in 8.12s ===========================================================================
(ptca) root@629bedab946c:/opt/pytorch# 
```

titaiwangms replied:
Could you open an issue and assign to me? I found the issue in the shape inference pass that doesn't provide static shape as it should. I suggest we have a more specific skip reason.

thiagocrepaldi replied:
Awesome, I was going to work on it myself. Here it is and thank you for being proactive
https://github.com/pytorch/pytorch/issues/105490

titaiwangms replied:
Thanks!

Commit:
Update on "Allow (temporarily?) non-fake input during ONNX export with fake mode"


Although input and model are expected to be fake during ONNX export with fake mode enabled, apparently some models can create new parameters during tracing. That makes internal checks on dynamo side to fail when we dont set `allow_non_fake_input=True` for `torch._dynamo.export`.

https://github.com/pytorch/pytorch/issues/105077 tracks this issue and if a proper fix is done, we will set `allow_non_fake_input=False` again

Additionally to that, a possible bug was found at torch.nn.Module.state_dict() in which some registered buffers are not listed.

This is being tracked by https://github.com/pytorch/pytorch/issues/105233 but in the mean time, we are merging `state_dict()` and `named_buffers()` results to create a full `state_dict` for the model

Two more complex/larger tests are added to the ONNX export which are the same for the experimental symbolic tracing: tiny gpt2 and toy mlp (https://github.com/pytorch/pytorch/blob/main/test/onnx/test_fx_to_onnx_with_onnxruntime.py#L766-L825)

ps: https://github.com/pytorch/pytorch/issues/105464 tracks pending tasks/limitations from this PR

[ghstack-poisoned]

Commit:
Update on "Allow (temporarily?) non-fake input during ONNX export with fake mode"


Although input and model are expected to be fake during ONNX export with fake mode enabled, apparently some models can create new parameters during tracing. That makes internal checks on dynamo side to fail when we dont set `allow_non_fake_input=True` for `torch._dynamo.export`.

https://github.com/pytorch/pytorch/issues/105077 tracks this issue and if a proper fix is done, we will set `allow_non_fake_input=False` again

Additionally to that, a possible bug was found at torch.nn.Module.state_dict() in which some registered buffers are not listed.

This is being tracked by https://github.com/pytorch/pytorch/issues/105233 but in the mean time, we are merging `state_dict()` and `named_buffers()` results to create a full `state_dict` for the model

Two more complex/larger tests are added to the ONNX export which are the same for the experimental symbolic tracing: tiny gpt2 and toy mlp (https://github.com/pytorch/pytorch/blob/main/test/onnx/test_fx_to_onnx_with_onnxruntime.py#L766-L825)

ps: https://github.com/pytorch/pytorch/issues/105464 tracks pending tasks/limitations from this PR

[ghstack-poisoned]

BowenBao commented on torch/onnx/_internal/fx/dynamo_graph_extractor.py:
```diff
@@ -194,6 +194,8 @@ def generate_fx(
             *model_args,
             tracing_mode=fx_mode,
             fake_mode=fake_mode,  # type: ignore[arg-type]
+            aten_graph=fake_mode
```
This is very high priority. Let's call soon.

thiagocrepaldi replied:
Sure. This is needed because of the integration between fake tensor and dynamic shape. Tracking through https://github.com/pytorch/pytorch/issues/105467 for now

thiagocrepaldi commented:
@pytorchbot merge
